File: ffupdater/src/main/java/de/marmaro/krt/ffupdater/security/StrictModeSetup.java
package de.marmaro.krt.ffupdater.security;

import android.os.Build;
import android.os.StrictMode;
import android.util.Log;

import de.marmaro.krt.ffupdater.BuildConfig;

/**
 * Configure StrictMode to improve security (by prohibit unencrypted network traffic) and detect bugs during development.
 */
public class StrictModeSetup {
    private static final String LOG_TAG = "StrictModeSetup";

    /**
     * If the app has been built locally, then StrictMode will be configured very repressive. This is necessary to find bugs in development fast.
     * If the app has been built by F-Droid, then forbid only unencrypted network traffic.
     */
    public static void enable() {
        if (BuildConfig.DEBUG) {
            enableDebugStrictMode();
        } else {
            enableReleaseStrictMode();
        }
    }

    private static void enableDebugStrictMode() {
        Log.i(LOG_TAG, "enable StrictMode for local development");
        StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder()
                .detectAll()
                .permitDiskReads() // for preferences
                .permitDiskWrites() // for update
                .permitNetwork() // for checking updates
                .penaltyLog()
                .penaltyDeath()
                .build());

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder()
                    .detectAll()
                    .penaltyLog()
                    .penaltyDeath()
                    .build());
        }
    }

    private static void enableReleaseStrictMode() {
        Log.i("MainActivity", "enable StrictMode for everyday usage to prevent unencrypted data connection");
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder()
                    .penaltyDeathOnCleartextNetwork()
                    .build());
        }
    }
}

File: ffupdater/src/main/java/de/marmaro/krt/ffupdater/security/StrictModeSetup.kt
package de.marmaro.krt.ffupdater.security

import android.os.StrictMode
import android.os.StrictMode.ThreadPolicy
import android.os.StrictMode.VmPolicy
import androidx.annotation.Keep
import de.marmaro.krt.ffupdater.BuildConfig
import de.marmaro.krt.ffupdater.device.DeviceSdkTester

/**
 * Configure StrictMode to find bugs or other problems.
 */
@Keep
object StrictModeSetup {
    fun enableStrictMode() {
        if (BuildConfig.DEBUG) {
            enableStrictModeForDebugging()
        } else {
            enableStrictModeForRelease()
        }
    }

    private fun enableStrictModeForDebugging() {
        StrictMode.setThreadPolicy(
            ThreadPolicy.Builder()
                .detectAll()
                .permitDiskReads() // for storing preferences
                .permitDiskWrites() // for downloading apps which will be installed
                .penaltyLog()
                .penaltyDialog()
                .build()
        )

        val vmPolicy = VmPolicy.Builder()
        vmPolicy.detectActivityLeaks()
        if (DeviceSdkTester.supportsAndroid6M23()) vmPolicy.detectCleartextNetwork()
        if (DeviceSdkTester.supportsAndroid8Oreo26()) vmPolicy.detectContentUriWithoutPermission()
        if (DeviceSdkTester.supportsAndroid10Q29()) vmPolicy.detectCredentialProtectedWhileLocked()
        vmPolicy.detectFileUriExposure()
        if (DeviceSdkTester.supportsAndroid10Q29()) vmPolicy.detectImplicitDirectBoot()
        if (DeviceSdkTester.supportsAndroid12S31()) vmPolicy.detectIncorrectContextUse()
        vmPolicy.detectLeakedClosableObjects()
        vmPolicy.detectLeakedRegistrationObjects()
        vmPolicy.detectLeakedSqlLiteObjects()
        // because of https://stackoverflow.com/a/53736775
//        if (DeviceSdkTester.supportsAndroid9()) vmPolicy.detectNonSdkApiUsage()
        if (DeviceSdkTester.supportsAndroid12S31()) vmPolicy.detectUnsafeIntentLaunch()
        if (DeviceSdkTester.supportsAndroid6M23()) vmPolicy.penaltyDeathOnCleartextNetwork()

        StrictMode.setVmPolicy(
            vmPolicy.build()
        )
    }

    private fun enableStrictModeForRelease() {
        if (!DeviceSdkTester.supportsAndroid6M23()) {
            return
        }
        val vmPolicy = VmPolicy.Builder()
            .penaltyDeathOnCleartextNetwork()
            .build()
        StrictMode.setVmPolicy(vmPolicy)
    }
}